File: flutter-boilerplate/lib/api/server/posts/stats.dart
import 'package:dio/dio.dart';
import 'package:flutter_boilerplate/lib/api_client.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

class PostStats {
  final int totalPosts;
  final int totalReactions;
  final double avgReactionsPerPost;

  const PostStats({
    required this.totalPosts,
    required this.totalReactions,
    required this.avgReactionsPerPost,
  });

  factory PostStats.fromJson(Map<String, dynamic> json) {
    return PostStats(
      totalPosts: json['totalPosts'] as int,
      totalReactions: json['totalReactions'] as int,
      avgReactionsPerPost: (json['avgReactionsPerPost'] as num).toDouble(),
    );
  }
}

final postStatsServerProvider =
    Provider((ref) => PostStatsServer(ref.read(dioProvider)));

const _query = 'query MyPostStats { myPostStats { totalPosts totalReactions avgReactionsPerPost } }';

class PostStatsServer {
  final Dio _dio;

  PostStatsServer(this._dio);

  Future<PostStats> call() async {
    final response = await _dio.post<dynamic>(
      '/graphql',
      data: {'query': _query},
    );
    final body = response.data as Map<String, dynamic>;
    if (body['errors'] != null) {
      throw DioException(
        requestOptions: response.requestOptions,
        message: 'Failed to fetch post stats',
      );
    }
    final result = (body['data'] as Map<String, dynamic>)['myPostStats']
        as Map<String, dynamic>;
    return PostStats.fromJson(result);
  }
}

File: tanstack-boilerplate/src/api/server/posts/stats.ts
import { apiFetch } from "@/lib/api-client";
import { GQL_URL } from "@/constants/api/urls";
import { POST } from "@/constants/api/methods";
import { JSON_CONTENT_TYPE_HEADER } from "@/constants/api/headers";

export interface PostStats {
  totalPosts: number;
  totalReactions: number;
  avgReactionsPerPost: number;
}

export async function fetchPostStatsServer(): Promise<PostStats> {
  const res = await apiFetch(GQL_URL, {
    method: POST,
    headers: JSON_CONTENT_TYPE_HEADER,
    body: JSON.stringify({
      query: `query { myPostStats { totalPosts totalReactions avgReactionsPerPost } }`,
    }),
  });
  if (!res.ok) {
    const data = await res.json().catch(() => ({}));
    throw new Error(data.error ?? "Failed to load stats");
  }
  const data = await res.json();
  if (!data.data?.myPostStats) throw new Error("No stats data returned");
  return data.data.myPostStats as PostStats;
}
